### ISSUE TITLE
RecurrenceEnter method fails because of FilterLimit and can not be changed (even with DI)

### ISSUE THREAD

**Issue by se2tka** [OData/odata.net#2700]
RecurrenceEnter method fails because of FilterLimit and can not be changed (even with DI).
I tried to change this limit by adding ODataUriParserSettings to dependency injection, but it was not used
I use this library through Microsoft.AspNetCore.OData.

### Expected result

It does not crash or limit can be changed.

**Comment by habbes** [comment]
Hello @se2tka. What is `RecurrenceEnter`? Where is it defined? Is it part of `Microsoft.AspNetCore.OData`? Could you share more details about your configuration, snippets of the code that's failing and sample requests? Or ideally, could you share a sample project that reproduces the issue?

**Comment by se2tka** [comment]
> Hello @se2tka. What is `RecurrenceEnter`? Where is it defined? Is it part of `Microsoft.AspNetCore.OData`? Could you share more details about your configuration, snippets of the code that's failing and sample requests? Or ideally, could you share a sample project that reproduces the issue?

RecurrenceEnter is a method from Microsoft.OData.Core library
Microsoft.OData.UriParser.BindingState class RecurseEnter method.

Problem is that I am having a UI (external library) that based on user input generates quite a complex query to Microsoft.AspNetCore.OData and then to Microsoft.OData.Core library

And Microsoft.OData.Core eventually crashes because of recursion limit in this RecurrenceEnter method.
An ideal solution for my project would be to increase FilterLimit value that RecurrenceEnter is using.
But it is quite hard from an external project as far as I see from your code.

**Comment by habbes** [comment]
@se2tka I see, you're referring to the `RecurseEnter` method in the internal `BinderState` class, right? Here's the implementation for reference:

```c#
        internal void RecurseEnter()
        {
            this.BindingRecursionDepth++;

            // TODO: add BindingLimit, use uniform error message
            if (this.BindingRecursionDepth > this.configuration.Settings.FilterLimit)
            {
                throw new ODataException(ODataErrorStrings.UriQueryExpressionParser_TooDeep);
            }
        }
```

As you have pointed out, the `FilterLimit` property is internal. I'm not sure why this is the case. I can check whether there's a way to override it from some public API.

However, the default limit on the recursion depth is 800. This is quite a high number. I'm more interested in learning about the use-case that exceeds this limit. Could the query be reconstructed such that it does not exceed this limit? At the end of the day, even if we remove this limit, there's a platform limit of how large the stack can grow, beyond which you'll get a stackoverflow exception. So, it may be a good idea to consider whether your use case can be achieved within limits.

**Comment by se2tka** [comment]
Thank you for your answer. Yes we will consider some workarounds.

The use case is that we have a WPF Telerik GridView control that is directly bound to an OData endpoint.
It has a UI feature to do filtering on a column with showing distinct values.
Some columns have up to 2000 distinct values.
And when a user sometimes mark all distinct values and uncheck some of them - it can make a query with a big IN statement (in MS SQL Server terms).

Users are not really aware of that and if you ask them they are very happy to see all distinct values and are happy to do this :)

In the other part of application (before we tried OData) we made handmade solution that does very similar thing.
It required a lot of work for that one grid, but actually it works. And SQL Server can handle such crazy query without a problem.
